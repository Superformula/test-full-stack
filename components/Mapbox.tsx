import { ReactElement, LegacyRef } from 'react'

import styles from './Mapbox.module.css'

interface Props {
  forwardRef: LegacyRef<HTMLDivElement>;
}

export default function Mapbox({ forwardRef }: Props): ReactElement {
  return <div className={styles.map} ref={forwardRef} />
}
