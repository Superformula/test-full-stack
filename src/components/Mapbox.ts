import styled from "@emotion/styled";
import { captureException } from "@sentry/browser";
import mapboxgl, { LngLatBoundsLike, LngLatLike } from "mapbox-gl";
import * as React from "react";

const ACCESS_TOKEN =
  "pk.eyJ1IjoiZ3p0b21hcyIsImEiOiJja2h6OGZuczkwaXNzMnNsMmw3dnk5bmxzIn0.QO9AD8ynVhAuNnjCKGLzVw";
const MAPBOX_STYLE = "mapbox://styles/gztomas/cki0qq81t40f619n23yfhor92";

export const geocode = async (query: string) => {
  const uri = `https://api.mapbox.com/geocoding/v5/mapbox.places/${query}.json?access_token=${ACCESS_TOKEN}`;
  const result = await fetch(uri);
  return (await result.json()) as {
    features: ({ bbox: LngLatBoundsLike; center: LngLatLike } | undefined)[];
  };
};

export const Mapbox = styled.div`
  border-radius: 8px;
  /* 
   * This specific color is not part of the theme but matches mapbox loading
   * color
   */
  background-color: #ece1cb;
  flex: 1;
  overflow: hidden;
  position: relative;

  canvas {
    outline: none;
  }
  .mapboxgl-ctrl-attrib {
    display: none;
  }
`;

export const useMapbox = (query: string | null) => {
  const ref = React.useRef<HTMLDivElement>(null);
  const mapboxRef = React.useRef<mapboxgl.Map>();
  const markerRef = React.useRef<mapboxgl.Marker>();

  React.useEffect(() => {
    const update = async () => {
      try {
        if (!query) return;
        const location = (await geocode(query)).features[0];
        if (location) {
          const element = ref.current;
          if (element && !element.children.length) {
            const container = document.createElement("div");
            const { style } = container;
            // These styles are needed inline so mapbox can pick them
            style.position = "absolute";
            style.top = style.right = style.bottom = style.left = "0";
            element.appendChild(container);
            mapboxRef.current = new mapboxgl.Map({
              accessToken: ACCESS_TOKEN,
              container,
              style: MAPBOX_STYLE,
              center: location.center,
              zoom: 10,
            });
            markerRef.current = new mapboxgl.Marker()
              .setLngLat(location.center)
              .addTo(mapboxRef.current);
          }
          mapboxRef.current?.setCenter(location.center);
          markerRef.current?.setLngLat(location.center);
          if (location.bbox) {
            mapboxRef.current?.fitBounds(location.bbox);
          } else {
            mapboxRef.current?.setZoom(13);
          }
        }
      } catch (e) {
        captureException(e, { extra: { query } });
      }
    };

    const timeout = setTimeout(() => void update(), 500);
    return () => clearTimeout(timeout);
  }, [query]);

  return ref;
};
