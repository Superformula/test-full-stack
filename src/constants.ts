export const MAPBOX_ACCESS_TOKEN =
  "pk.eyJ1IjoiZ3p0b21hcyIsImEiOiJja2h6OGZuczkwaXNzMnNsMmw3dnk5bmxzIn0.QO9AD8ynVhAuNnjCKGLzVw";
export const MAPBOX_STYLE = "mapbox://styles/gztomas/cki0qq81t40f619n23yfhor92";
export const PLACEHOLDER_AVATAR_IMAGE =
  "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";
