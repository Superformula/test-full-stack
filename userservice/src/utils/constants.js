const Constants = {
  ADD_USER: "AddUser",
  GET_USER: "GetUser",
  GET_USERS: "GetUsers",
  UPDATE_USER: "UpdateUser",
  DELETE_USER: "DeleteUser",
};

export default Constants;
