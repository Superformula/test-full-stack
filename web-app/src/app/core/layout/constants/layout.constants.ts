import { InjectionToken } from '@angular/core';

export const LAYOUT_CONFIG = new InjectionToken<any>('Default configuration for the app');

