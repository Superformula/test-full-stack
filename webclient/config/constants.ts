export const siteMetadata = {
  title: 'Users list'
}

export const DEFAULT_PAGE_SIZE = 6
