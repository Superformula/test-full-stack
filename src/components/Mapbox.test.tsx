import { captureException } from "@sentry/browser";
import { render, waitFor } from "@testing-library/react";
import mapboxgl from "mapbox-gl";
import * as React from "react";
import { MAPBOX_ACCESS_TOKEN } from "../constants";
import { useMapbox } from "./Mapbox";

jest.mock("@sentry/browser");
jest.useFakeTimers();

const fetchMock = jest.fn(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve({
        features: [{ bbox: [1, 2, 3, 4], center: [1, 2] }],
      }),
  } as Response)
);

test("useMapbox", async () => {
  window.fetch = fetchMock;

  const Component = () => {
    const ref = useMapbox("London");
    return <div ref={ref}>map here</div>;
  };

  const { container } = render(<Component />);

  expect(fetchMock).not.toHaveBeenCalled();
  expect(mapboxgl.Map).not.toHaveBeenCalledWith();
  expect(mapboxgl.Marker).not.toHaveBeenCalledWith();

  jest.runAllTimers();

  expect(captureException).not.toHaveBeenCalled();
  expect(fetchMock).toHaveBeenCalledWith(
    `https://api.mapbox.com/geocoding/v5/mapbox.places/London.json?access_token=${MAPBOX_ACCESS_TOKEN}`
  );

  // Promises need to be resolved
  await waitFor(() => {
    expect(mapboxgl.Map).toHaveBeenCalledWith(
      expect.objectContaining({
        accessToken: MAPBOX_ACCESS_TOKEN,
        center: [1, 2],
        style: "mapbox://styles/gztomas/cki0qq81t40f619n23yfhor92",
        zoom: 10,
      })
    );
    expect(mapboxgl.Marker).toHaveBeenCalled();
  });

  expect(container).toMatchSnapshot();
});
