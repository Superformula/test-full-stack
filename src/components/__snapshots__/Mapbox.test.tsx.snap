// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMapbox 1`] = `
<div>
  <div>
    map here
    <div
      style="position: absolute; left: 0px; bottom: 0px; right: 0px; top: 0px;"
    />
  </div>
</div>
`;
